From d1f70a986aead317d03179b31111055f8eb51a96 Mon Sep 17 00:00:00 2001
Message-Id: <d1f70a986aead317d03179b31111055f8eb51a96.1711648979.git.keller@fs-net.de>
In-Reply-To: <bb92f3f011b7b125a61ac948b576f518cee03bb3.1711648979.git.keller@fs-net.de>
References: <bb92f3f011b7b125a61ac948b576f518cee03bb3.1711648979.git.keller@fs-net.de>
From: Tiger Yu <tfyu@codeaurora.org>
Date: Wed, 17 Feb 2021 13:15:59 +0800
Subject: [PATCH 13/16] qcacld-2.0: Drop mcast and plaintext frags in protected
 network

Multicast frames should not be fragmented and plaintext frags should not
be reassembeld in protected network.

Fix is to drop mcast frags and plaintext frags received in protected
network.

Change-Id: Ibed9015b27e82e311ce7a019ff0047b123e5b2d8
CRs-Fixed: 2868034
---
 CORE/CLD_TXRX/TXRX/ol_rx_defrag.c | 33 +++++++++++++++++++++++++++++++++
 1 file changed, 33 insertions(+)

diff --git a/CORE/CLD_TXRX/TXRX/ol_rx_defrag.c b/CORE/CLD_TXRX/TXRX/ol_rx_defrag.c
index e6748c9..014a3a7 100644
--- a/CORE/CLD_TXRX/TXRX/ol_rx_defrag.c
+++ b/CORE/CLD_TXRX/TXRX/ol_rx_defrag.c
@@ -283,6 +283,8 @@ ol_rx_reorder_store_frag(
     struct ol_rx_reorder_array_elem_t *rx_reorder_array_elem;
     u_int16_t frxseq, rxseq, seq;
     htt_pdev_handle htt_pdev = pdev->htt_pdev;
+    void *rx_desc;
+    int index;              /* unicast vs. multicast */
 
     seq = seq_num & peer->tids_rx_reorder[tid].win_sz_mask;
     adf_os_assert(seq == 0);
@@ -295,6 +297,37 @@ ol_rx_reorder_store_frag(
         IEEE80211_SEQ_FRAG_MASK;
     more_frag = mac_hdr->i_fc[1] & IEEE80211_FC1_MORE_FRAG;
 
+    /*
+     * CR 2868034
+     * A broadcast or multicast frame should never be fragmented
+     * driver should drop all fragmented broadcast / multicast frames.
+     */
+    rx_desc = htt_rx_msdu_desc_retrieve(htt_pdev, frag);
+    if (htt_rx_msdu_is_wlan_mcast(htt_pdev, rx_desc)) {
+        /* Discard Mcast/Bcast single frag frames */
+        ol_rx_frames_free(htt_pdev, frag);
+
+        VOS_TRACE(VOS_MODULE_ID_TXRX, VOS_TRACE_LEVEL_ERROR,
+            "%s: QSV2020004 discard MCAST/BCAST frag frame\n\n",
+            __FUNCTION__);
+        return;
+    }
+
+    index = htt_rx_msdu_is_wlan_mcast(htt_pdev, rx_desc) ?
+                txrx_sec_mcast : txrx_sec_ucast;
+
+    if (peer->security[index].sec_type != htt_sec_type_none &&
+        !htt_rx_mpdu_is_encrypted(htt_pdev, rx_desc)) {
+        ol_rx_frames_free(htt_pdev, frag);
+
+        VOS_TRACE(VOS_MODULE_ID_TXRX, VOS_TRACE_LEVEL_ERROR,
+            "%s: QSV2020004, Unencrypted fragment received in security mode %d\n",
+            __FUNCTION__,
+            peer->security[index].sec_type);
+
+        return;
+    }
+
     if ((!more_frag) && (!fragno) && (!rx_reorder_array_elem->head)) {
         ol_rx_fraglist_insert(htt_pdev, peer,
             &rx_reorder_array_elem->head,
-- 
2.14.5

